rpc/export.go

```diff
@@ -154,26 +155,71 @@ func (c *lockedConn) sendCap(d rpccp.CapDescriptor, client capnp.Client) (_ expo
 	state.Metadata.Lock()
 	defer state.Metadata.Unlock()
 	id, ok := c.findExportID(state.Metadata)
+	var ee *expent
 	if ok {
-		ent := c.lk.exports[id]
-		ent.wireRefs++
-		d.SetSenderHosted(uint32(id))
-		return id, true, nil
+		ee = c.lk.exports[id]
+		ee.wireRefs++
+	} else {
+		// Not already present; allocate an export id for it:
+		ee = &expent{
+			client:    client.AddRef(),
+			wireRefs:  1,
+			isPromise: state.IsPromise,
+		}
+		id = exportID(c.lk.exportID.next())
+		if int64(id) == int64(len(c.lk.exports)) {
+			c.lk.exports = append(c.lk.exports, ee)
+		} else {
+			c.lk.exports[id] = ee
+		}
+		c.setExportID(state.Metadata, id)
 	}
+	if ee.isPromise {
+		// Send a promise, wait for the resolution asynchronously, then send
+		// a resolve message:
+		d.SetSenderPromise(uint32(id))
+		waitRef := client.AddRef()
+		go func() {
+			unlockedConn := (*Conn)(c)
```

Comment:
Do you want to assert that the conn is actually unlocked?  As I recall, you had a method that would panic if that wasn't the case.

No need if you're confident it's always unlocked.

Reply:
I'm not sure what method your thinking of, or how it would even work.

In this case, this is at the top of a new goroutine, so it logically doesn't "own" c; when it needs the lock it should re-acquire it. It's a bit annoying that we can't get the unlocked version without a cast. Maybe this could use a comment.

Reply:
Update: added a comment.